src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -0,0 +1,356 @@
+                detach=True,
+                network_mode="host",
+                environment=self.env,
+                volumes=self.vol,
+            )
+        elif status == "existed":
+            container.start()
+        elif status == "paused":
+            container.unpause()
+        else:
+            print("container is restarting or running ...")
+        self.container = container
+
+    def stop_callback_server(self):
+        self.container.stop()
+        self.container.remove()
+
+
+def _make_output_dir(root_dir, source_data_dir, source, workspace_id):
+    """Helper function that makes output directory for a specific collection under root directory."""
+    output_dir = os.path.join(root_dir, source_data_dir, source, str(workspace_id))
+    os.makedirs(output_dir, exist_ok=True)
+    return output_dir
+
+
+def _make_job_dir(root_dir, job_dir, username):
+    """Helper function that create a job_dir for a user under root directory."""
+    job_dir = os.path.join(root_dir, job_dir, username)
+    os.makedirs(job_dir, exist_ok=True)
+    # only user can cread, write, or execute
+    os.chmod(job_dir, stat.S_IRWXU)
+    return job_dir
+
+
+def _list_objects_params(wsid, min_id, max_id, type_str):
+    """Helper function that creats params needed for list_objects function."""
+    params = {
+        "ids": [wsid],
+        "minObjectID": min_id,
+        "maxObjectID": max_id,
+        "type": type_str,
+    }
+    return params
+
+
+def _process_object_info(obj_info):
+    """
+    "upa", "name", "type", and "timestamp info will be extracted from object info and save as a dict."
+    {
+        "upa": "790541/67/2",
+        "name": <copy object name from object info>
+        "type": <copy object type from object info>
+        "timestamp": <copy timestamp from object info>
+    }
+    """
+    res_dict = {}
+    res_dict["upa"] = "{6}/{0}/{4}".format(*obj_info)
+    res_dict["name"] = obj_info[1]
+    res_dict["type"] = obj_info[2]
+    res_dict["timestamp"] = obj_info[3]
+    return res_dict
+
+
+def list_objects(wsid, conf, filter_objects_name_by, batch_size=10000):
+    """
+    List all objects information given a workspace ID.
+    """
+    if batch_size > 10000:
+        raise ValueError("Maximum value for listing workspace objects is 10000")
+
+    maxObjectID = conf.ws.get_workspace_info({"id": wsid})[4]
+    batch_input = [
+        [idx + 1, idx + batch_size] for idx in range(0, maxObjectID, batch_size)
+    ]
+    objs = [
+        conf.ws.list_objects(
+            _list_objects_params(wsid, min_id, max_id, filter_objects_name_by)
+        )
+        for min_id, max_id in batch_input
+    ]
+    res_objs = list(itertools.chain.from_iterable(objs))
+    return res_objs
+
+
+def process_input(conf):
+    """
+    Download .fa and .meta files from workspace and save a copy under output_dir.
+    """
+    while True:
+        task = conf.queue.get(block=True)
+        if not task:
+            print("Stopping")
```

Comment:
Optional, but this would be more helpful if it printed the process ID or something as well